File: web_app/static/js/lineage_analyzer.js
class LineageAnalyzer {
    constructor() {
        this.currentChart = null;
        this.currentLineage = null;
        this.searchTimeout = null;
        
        this.initializeEventListeners();
        this.loadInitialData();
    }
    
    initializeEventListeners() {
        const searchInput = document.getElementById('lineageSearch');
        const searchResults = document.getElementById('searchResults');
        
        // Search functionality
        searchInput.addEventListener('input', (e) => {
            clearTimeout(this.searchTimeout);
            this.searchTimeout = setTimeout(() => {
                this.performSearch(e.target.value);
            }, 300);
        });
        
        // Hide search results when clicking outside
        document.addEventListener('click', (e) => {
            if (!searchInput.contains(e.target) && !searchResults.contains(e.target)) {
                searchResults.style.display = 'none';
            }
        });
        
        // Show search results when focusing on input
        searchInput.addEventListener('focus', () => {
            if (searchResults.children.length > 0) {
                searchResults.style.display = 'block';
            }
        });
    }
    
    async loadInitialData() {
        // Load top accelerators by default
        await this.performSearch('');
    }
    
    async performSearch(query) {
        try {
            const response = await fetch(`/api/lineage/search?query=${encodeURIComponent(query)}`);
            const data = await response.json();
            
            this.displaySearchResults(data.results || []);
        } catch (error) {
            console.error('Search error:', error);
            this.showError('Error performing search: ' + error.message);
        }
    }
    
    displaySearchResults(results) {
        const searchResults = document.getElementById('searchResults');
        
        if (results.length === 0) {
            searchResults.innerHTML = '<div class="search-result-item">No results found</div>';
            searchResults.style.display = 'block';
            return;
        }
        
        searchResults.innerHTML = '';
        
        results.forEach(result => {
            const item = document.createElement('div');
            item.className = 'search-result-item';
            item.onclick = () => this.selectLineage(result.lineage_id);
            
            const accelerationClass = this.getAccelerationClass(result.acceleration_pct_per_year);
            
            item.innerHTML = `
                <div class="result-org">${result.parent_organization}</div>
                <div class="result-metrics">
                    ${result.first_year}-${result.last_year} • ${result.years_tracked} years • ${result.latest_enrollment.toLocaleString()} enrollees
                    <span class="result-acceleration ${accelerationClass}">
                        ${result.acceleration_pct_per_year > 0 ? '+' : ''}${result.acceleration_pct_per_year}%/yr
                    </span>
                </div>
            `;
            
            searchResults.appendChild(item);
        });
        
        searchResults.style.display = 'block';
    }
    
    getAccelerationClass(acceleration) {
        if (acceleration >= 5) return 'acceleration-high';
        if (acceleration >= 2) return 'acceleration-medium';
        if (acceleration >= -2) return 'acceleration-low';
        return 'acceleration-negative';
    }
    
    async selectLineage(lineageId) {
        try {
            // Hide search results
            document.getElementById('searchResults').style.display = 'none';
            
            // Show loading state
            this.showLoadingState();
            
            // Fetch detailed lineage data
            const response = await fetch(`/api/lineage/lineage/${lineageId}`);
            if (!response.ok) {
                throw new Error(`HTTP ${response.status}: ${response.statusText}`);
            }
            
            const data = await response.json();
            this.currentLineage = data;
            
            // Display the lineage details
            this.displayLineageDetails(data);
            this.createLineageChart(data);
            this.displayYearlyData(data.yearly_data);
            this.showCalculationExplanation();
            
        } catch (error) {
            console.error('Error loading lineage:', error);
            this.showError('Error loading lineage details: ' + error.message);
        }
    }
    
    showLoadingState() {
        document.getElementById('lineageDetails').style.display = 'grid';
        document.getElementById('lineageTitle').textContent = 'Loading...';
        document.getElementById('lineageSubtitle').textContent = 'Fetching lineage details...';
    }
    
    displayLineageDetails(data) {
        const summary = data.summary;
        
        // Update title and subtitle
        document.getElementById('lineageTitle').textContent = summary.parent_organization;
        document.getElementById('lineageSubtitle').textContent = 
            `${summary.plan_type} • ${summary.first_year}-${summary.last_year} • ID: ${summary.lineage_id}`;
        
        // Update acceleration badge
        const accelerationBadge = document.getElementById('accelerationBadge');
        const acceleration = summary.acceleration_vs_competitors_pct_per_year;
        accelerationBadge.textContent = `${acceleration > 0 ? '+' : ''}${acceleration.toFixed(2)}%/year`;
        accelerationBadge.className = `acceleration-badge ${this.getAccelerationClass(acceleration)}`;
        
        // Update metrics
        document.getElementById('yearsTracked').textContent = summary.years_tracked;
        document.getElementById('latestEnrollment').textContent = summary.latest_enrollment.toLocaleString();
        document.getElementById('rSquared').textContent = summary.r_squared_competitors.toFixed(3);
        
        const performance = summary.current_performance_vs_competitors_pct;
        const performanceElement = document.getElementById('currentPerformance');
        performanceElement.textContent = `${performance > 0 ? '+' : ''}${performance.toFixed(1)}%`;
        performanceElement.className = `metric-value ${performance > 0 ? 'performance-positive' : 'performance-negative'}`;
        
        // Show the details
        document.getElementById('lineageDetails').style.display = 'grid';
    }
    
    createLineageChart(data) {
        const ctx = document.getElementById('lineageChart').getContext('2d');
        
        // Destroy existing chart if any
        if (this.currentChart) {
            this.currentChart.destroy();
        }
        
        const graphData = data.graph_data;
        const trendData = data.trend_data;
        
        // Prepare datasets
        const datasets = [
            {
                label: 'Actual Risk Score',
                data: graphData.years.map((year, i) => ({
                    x: year,
                    y: graphData.actual_risk_scores[i]
                })).filter(d => d.y !== null),
                borderColor: '#e74c3c',
                backgroundColor: 'rgba(231, 76, 60, 0.1)',
                borderWidth: 3,
                pointRadius: 6,
                pointHoverRadius: 8,
                tension: 0.2
            },
            {
                label: 'Competitor Benchmark',
                data: graphData.years.map((year, i) => ({
                    x: year,
                    y: graphData.benchmark_scores[i]
                })).filter(d => d.y !== null),
                borderColor: '#3498db',
                backgroundColor: 'rgba(52, 152, 219, 0.1)',
                borderWidth: 3,
                pointRadius: 6,
                pointHoverRadius: 8,
                tension: 0.2
            }
        ];
        
        // Add trend lines if available
        if (trendData.actual_trend) {
            datasets.push({
                label: 'Actual Trend',
                data: graphData.years.slice(0, trendData.actual_trend.values.length).map((year, i) => ({
                    x: year,
                    y: trendData.actual_trend.values[i]
                })),
                borderColor: '#c0392b',
                borderWidth: 2,
                borderDash: [5, 5],
                pointRadius: 0,
                tension: 0
            });
        }
        
        if (trendData.benchmark_trend) {
            datasets.push({
                label: 'Benchmark Trend',
                data: graphData.years.slice(0, trendData.benchmark_trend.values.length).map((year, i) => ({
                    x: year,
                    y: trendData.benchmark_trend.values[i]
                })),
                borderColor: '#2980b9',
                borderWidth: 2,
                borderDash: [5, 5],
                pointRadius: 0,
                tension: 0
            });
        }
        
        this.currentChart = new Chart(ctx, {
            type: 'line',
            data: { datasets },
            options: {
                responsive: true,
                maintainAspectRatio: false,
                interaction: {
                    intersect: false,
                    mode: 'index'
                },
                plugins: {
                    title: {
                        display: true,
                        text: 'Risk Score Acceleration Analysis',
                        font: { size: 16, weight: 'bold' }
                    },
                    legend: {
                        position: 'top'
                    },
                    tooltip: {
                        callbacks: {
                            afterBody: (context) => {
                                const dataIndex = context[0].dataIndex;
                                const enrollment = graphData.enrollment_sizes[dataIndex];
                                const performance = graphData.performance_vs_competitors[dataIndex];
                                
                                let extra = [];
                                if (enrollment) extra.push(`Enrollment: ${enrollment.toLocaleString()}`);
                                if (performance !== null) extra.push(`vs Competitors: ${performance > 0 ? '+' : ''}${performance.toFixed(2)}%`);
                                
                                return extra;
                            }
                        }
                    }
                },
                scales: {
                    x: {
                        type: 'linear',
                        position: 'bottom',
                        title: {
                            display: true,
                            text: 'Year'
                        },
                        ticks: {
                            stepSize: 1,
                            callback: function(value) {
                                return Math.floor(value).toString();
                            }
                        }
                    },
                    y: {
                        title: {
                            display: true,
                            text: 'Risk Score'
                        },
                        ticks: {
                            callback: function(value) {
                                return value.toFixed(3);
                            }
                        }
                    }
                }
            }
        });
    }
    
    displayYearlyData(yearlyData) {
        const tableBody = document.querySelector('#yearlyDataTable tbody');
        tableBody.innerHTML = '';
        
        yearlyData.forEach(year => {
            const row = document.createElement('tr');
            
            // Format performance with color coding
            const performance = year.performance_vs_market_excluding_pct;
            const performanceClass = performance > 0 ? 'performance-positive' : 'performance-negative';
            const performanceText = performance !== null ? 
                `${performance > 0 ? '+' : ''}${performance.toFixed(2)}%` : 'N/A';
            
            // Format counties list (truncate if too long)
            let countiesText = year.counties_in_footprint || 'N/A';
            if (countiesText.length > 50) {
                countiesText = countiesText.substring(0, 50) + '...';
            }
            
            row.innerHTML = `
                <td><strong>${year.year}</strong></td>
                <td><span class="plan-evolution">${year.contract_number}_${year.plan_id}</span></td>
                <td>${year.total_enrollment.toLocaleString()}</td>
                <td>${year.actual_risk_score !== null ? year.actual_risk_score.toFixed(4) : 'N/A'}</td>
                <td>${year.county_weighted_benchmark_excluding !== null ? year.county_weighted_benchmark_excluding.toFixed(4) : 'N/A'}</td>
                <td class="performance-cell ${performanceClass}">${performanceText}</td>
                <td title="${year.counties_in_footprint}">${year.num_counties} counties</td>
                <td>${year.total_footprint_enrollment.toLocaleString()}</td>
            `;
            
            tableBody.appendChild(row);
        });
        
        document.getElementById('yearlyDataSection').style.display = 'block';
    }
    
    showCalculationExplanation() {
        document.getElementById('calculationExplanation').style.display = 'block';
    }
    
    showError(message) {
        const container = document.querySelector('.lineage-container');
        const existingError = container.querySelector('.error');
        if (existingError) {
            existingError.remove();
        }
        
        const errorDiv = document.createElement('div');
        errorDiv.className = 'error';
        errorDiv.textContent = message;
        
        container.insertBefore(errorDiv, container.firstChild);
        
        // Auto-hide after 10 seconds
        setTimeout(() => {
            if (errorDiv.parentNode) {
                errorDiv.remove();
            }
        }, 10000);
    }
    
    // Utility function to format numbers
    formatNumber(num, decimals = 0) {
        if (num === null || num === undefined) return 'N/A';
        return num.toFixed(decimals).replace(/\B(?=(\d{3})+(?!\d))/g, ',');
    }
    
    // Utility function to format percentages
    formatPercentage(num, decimals = 2) {
        if (num === null || num === undefined) return 'N/A';
        const sign = num > 0 ? '+' : '';
        return `${sign}${num.toFixed(decimals)}%`;
    }
}

// Initialize the analyzer when the page loads
document.addEventListener('DOMContentLoaded', () => {
    window.lineageAnalyzer = new LineageAnalyzer();
});

// Export for potential external use
if (typeof module !== 'undefined' && module.exports) {
    module.exports = LineageAnalyzer;
} 

File: web_app/routers/lineage_analyzer.py
from fastapi import APIRouter, HTTPException
from fastapi.responses import JSONResponse
import pandas as pd
import numpy as np
from typing import Dict, List, Any, Optional
import os

router = APIRouter()

# Load the step-by-step analysis data
try:
    results_df = pd.read_csv('step_by_step_acceleration_results_20250703_130114.csv')
    lineage_df = pd.read_csv('step_by_step_lineage_data_20250703_130114.csv')
    print(f"Loaded lineage data: {len(results_df):,} lineages, {len(lineage_df):,} lineage-year records")
except Exception as e:
    print(f"Error loading lineage data: {e}")
    results_df = pd.DataFrame()
    lineage_df = pd.DataFrame()

@router.get("/search")
async def search_lineages(query: str = ""):
    """Search for lineages by organization name, contract number, or lineage ID"""
    
    if results_df.empty:
        raise HTTPException(status_code=500, detail="Lineage data not loaded")
    
    if not query or len(query) < 2:
        # Return top accelerators as default
        top_results = results_df.nlargest(20, 'acceleration_vs_competitors_pct_per_year')
    else:
        # Search in multiple fields
        query_lower = query.lower()
        mask = (
            results_df['parent_organization_name'].str.lower().str.contains(query_lower, na=False) |
            results_df['persistent_entity_id'].str.lower().str.contains(query_lower, na=False)
        )
        
        # Also search in lineage data for contract numbers
        if query_lower.startswith('h'):
            contract_mask = lineage_df['contract_number'].str.lower().str.contains(query_lower, na=False)
            matching_lineages = lineage_df[contract_mask]['persistent_entity_id'].unique()
            lineage_mask = results_df['persistent_entity_id'].isin(matching_lineages)
            mask = mask | lineage_mask
        
        top_results = results_df[mask].nlargest(50, 'acceleration_vs_competitors_pct_per_year')
    
    # Format results for frontend
    search_results = []
    for _, row in top_results.iterrows():
        search_results.append({
            'lineage_id': row['persistent_entity_id'],
            'parent_organization': row['parent_organization_name'],
            'plan_type': row['plan_type'],
            'acceleration_pct_per_year': round(row['acceleration_vs_competitors_pct_per_year'], 2),
            'r_squared': round(row['r_squared_competitors'], 3),
            'years_tracked': int(row['years_tracked']),
            'latest_enrollment': int(row['latest_enrollment']),
            'first_year': int(row['first_year']),
            'last_year': int(row['last_year'])
        })
    
    return {
        'results': search_results,
        'total_found': len(search_results)
    }

@router.get("/lineage/{lineage_id}")
async def get_lineage_details(lineage_id: str):
    """Get complete lineage details for a specific lineage ID"""
    
    if results_df.empty or lineage_df.empty:
        raise HTTPException(status_code=500, detail="Lineage data not loaded")
    
    # Get summary data
    summary_data = results_df[results_df['persistent_entity_id'] == lineage_id]
    if summary_data.empty:
        raise HTTPException(status_code=404, detail="Lineage not found")
    
    summary_row = summary_data.iloc[0]
    
    # Get year-by-year detailed data
    yearly_data = lineage_df[lineage_df['persistent_entity_id'] == lineage_id].sort_values('year')
    if yearly_data.empty:
        raise HTTPException(status_code=404, detail="No yearly data found for lineage")
    
    # Format summary metrics
    summary = {
        'lineage_id': lineage_id,
        'parent_organization': summary_row['parent_organization_name'],
        'plan_type': summary_row['plan_type'],
        'is_dual_eligible': bool(summary_row['is_dual_eligible']),
        'years_tracked': int(summary_row['years_tracked']),
        'first_year': int(summary_row['first_year']),
        'last_year': int(summary_row['last_year']),
        'latest_enrollment': int(summary_row['latest_enrollment']),
        'latest_risk_score': round(summary_row['latest_risk_score'], 4),
        'acceleration_vs_competitors_pct_per_year': round(summary_row['acceleration_vs_competitors_pct_per_year'], 3),
        'acceleration_vs_market_pct_per_year': round(summary_row['acceleration_vs_market_pct_per_year'], 3),
        'r_squared_competitors': round(summary_row['r_squared_competitors'], 4),
        'r_squared_market': round(summary_row['r_squared_market'], 4),
        'p_value_competitors': round(summary_row['p_value_competitors'], 6),
        'p_value_market': round(summary_row['p_value_market'], 6),
        'current_performance_vs_competitors_pct': round(summary_row['current_performance_vs_competitors_pct'], 2),
        'current_performance_vs_market_pct': round(summary_row['current_performance_vs_market_pct'], 2)
    }
    
    # Format yearly data
    yearly_details = []
    for _, row in yearly_data.iterrows():
        yearly_details.append({
            'year': int(row['year']),
            'contract_number': row['contract_number'],
            'plan_id': row['plan_id'],
            'parent_organization_name': row['parent_organization_name'],
            'plan_type': row['plan_type'],
            'is_dual_eligible': bool(row['is_dual_eligible']),
            'is_chronic_snp': bool(row['is_chronic_snp']),
            'is_institutional_snp': bool(row['is_institutional_snp']),
            'total_enrollment': int(row['total_enrollment']) if pd.notna(row['total_enrollment']) else 0,
            'actual_risk_score': round(row['actual_risk_score'], 4) if pd.notna(row['actual_risk_score']) else None,
            'num_counties': int(row['num_counties']) if pd.notna(row['num_counties']) else 0,
            'county_weighted_benchmark_including': round(row['county_weighted_benchmark_including'], 4) if pd.notna(row['county_weighted_benchmark_including']) else None,
            'county_weighted_benchmark_excluding': round(row['county_weighted_benchmark_excluding'], 4) if pd.notna(row['county_weighted_benchmark_excluding']) else None,
            'counties_in_footprint': row['counties_in_footprint'] if pd.notna(row['counties_in_footprint']) else '',
            'total_footprint_enrollment': int(row['total_footprint_enrollment']) if pd.notna(row['total_footprint_enrollment']) else 0,
            'performance_vs_market_including_pct': round(row['performance_vs_market_including_pct'], 3) if pd.notna(row['performance_vs_market_including_pct']) else None,
            'performance_vs_market_excluding_pct': round(row['performance_vs_market_excluding_pct'], 3) if pd.notna(row['performance_vs_market_excluding_pct']) else None
        })
    
    # Create graph data for time series
    graph_data = {
        'years': [d['year'] for d in yearly_details],
        'actual_risk_scores': [d['actual_risk_score'] for d in yearly_details],
        'benchmark_scores': [d['county_weighted_benchmark_excluding'] for d in yearly_details],
        'enrollment_sizes': [d['total_enrollment'] for d in yearly_details],
        'performance_vs_competitors': [d['performance_vs_market_excluding_pct'] for d in yearly_details]
    }
    
    # Calculate trend lines
    years_array = np.array(graph_data['years'])
    actual_scores = np.array([s for s in graph_data['actual_risk_scores'] if s is not None])
    benchmark_scores = np.array([s for s in graph_data['benchmark_scores'] if s is not None])
    
    trend_data = {}
    if len(actual_scores) >= 2:
        # Calculate trend line for actual scores
        valid_years = years_array[:len(actual_scores)]
        actual_trend = np.polyfit(valid_years, actual_scores, 1)
        trend_data['actual_trend'] = {
            'slope': float(actual_trend[0]),
            'intercept': float(actual_trend[1]),
            'values': [actual_trend[0] * year + actual_trend[1] for year in valid_years]
        }
        
        # Calculate trend line for benchmark
        if len(benchmark_scores) >= 2:
            benchmark_trend = np.polyfit(valid_years[:len(benchmark_scores)], benchmark_scores, 1)
            trend_data['benchmark_trend'] = {
                'slope': float(benchmark_trend[0]),
                'intercept': float(benchmark_trend[1]),
                'values': [benchmark_trend[0] * year + benchmark_trend[1] for year in valid_years[:len(benchmark_scores)]]
            }
    
    return {
        'summary': summary,
        'yearly_data': yearly_details,
        'graph_data': graph_data,
        'trend_data': trend_data
    } 